File: agent-app/src/test/kotlin/com/riwonace/agent/core/ModelEscalatorTest.kt
package com.riwonace.agent.core

import com.riwonace.agent.router.Route
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

class ModelEscalatorTest {

    private val escalator = ModelEscalator(
        enabled = true,
        smallModel = "gemma3:1b",
        mediumModel = "qwen2.5:3b",
        largeModel = "qwen2.5:7b",
    )

    @Test
    fun `낮은 복잡도는 SMALL 모델을 선택한다`() {
        val profile = createProfile(complexity = 0.2, uncertainty = 0.1)

        val selection = escalator.selectModel(profile)

        assertEquals(ModelEscalator.ModelTier.SMALL, selection.tier)
        assertEquals("gemma3:1b", selection.model)
    }

    @Test
    fun `중간 복잡도는 MEDIUM 모델을 선택한다`() {
        val profile = createProfile(complexity = 0.5, uncertainty = 0.1)

        val selection = escalator.selectModel(profile)

        assertEquals(ModelEscalator.ModelTier.MEDIUM, selection.tier)
        assertEquals("qwen2.5:3b", selection.model)
    }

    @Test
    fun `높은 복잡도는 LARGE 모델을 선택한다`() {
        val profile = createProfile(complexity = 0.8, uncertainty = 0.1)

        val selection = escalator.selectModel(profile)

        assertEquals(ModelEscalator.ModelTier.LARGE, selection.tier)
        assertEquals("qwen2.5:7b", selection.model)
    }

    @Test
    fun `높은 불확실성은 한 단계 상향한다`() {
        val profile = createProfile(complexity = 0.3, uncertainty = 0.6)

        val selection = escalator.selectModel(profile)

        assertEquals(ModelEscalator.ModelTier.MEDIUM, selection.tier) // SMALL → MEDIUM
    }

    @Test
    fun `다단계 추론은 최소 MEDIUM 모델이다`() {
        val profile = createProfile(complexity = 0.2, uncertainty = 0.1, isMultiHop = true)

        val selection = escalator.selectModel(profile)

        assertEquals(ModelEscalator.ModelTier.MEDIUM, selection.tier)
    }

    @Test
    fun `품질이 낮으면 재에스컬레이션을 권장한다`() {
        val currentSelection = ModelEscalator.ModelSelection(
            model = "gemma3:1b",
            tier = ModelEscalator.ModelTier.SMALL,
            reason = "테스트",
            latencyMultiplier = 1.0,
        )

        val reescalation = escalator.shouldReescalate(
            currentSelection,
            answerQuality = 0.4,
            claimCoverage = 0.5,
        )

        assertEquals(ModelEscalator.ModelTier.MEDIUM, reescalation?.tier)
    }

    @Test
    fun `품질이 충분하면 재에스컬레이션하지 않는다`() {
        val currentSelection = ModelEscalator.ModelSelection(
            model = "gemma3:1b",
            tier = ModelEscalator.ModelTier.SMALL,
            reason = "테스트",
            latencyMultiplier = 1.0,
        )

        val reescalation = escalator.shouldReescalate(
            currentSelection,
            answerQuality = 0.8,
            claimCoverage = 0.9,
        )

        assertEquals(null, reescalation)
    }

    @Test
    fun `이미 LARGE 모델이면 재에스컬레이션하지 않는다`() {
        val currentSelection = ModelEscalator.ModelSelection(
            model = "qwen2.5:7b",
            tier = ModelEscalator.ModelTier.LARGE,
            reason = "테스트",
            latencyMultiplier = 5.0,
        )

        val reescalation = escalator.shouldReescalate(
            currentSelection,
            answerQuality = 0.3,
            claimCoverage = 0.3,
        )

        assertEquals(null, reescalation)
    }

    @Test
    fun `에스컬레이션 비활성화 시 항상 SMALL 모델`() {
        val disabledEscalator = ModelEscalator(
            enabled = false,
            smallModel = "gemma3:1b",
            mediumModel = "qwen2.5:3b",
            largeModel = "qwen2.5:7b",
        )
        val profile = createProfile(complexity = 0.9, uncertainty = 0.9)

        val selection = disabledEscalator.selectModel(profile)

        assertEquals(ModelEscalator.ModelTier.SMALL, selection.tier)
    }

    private fun createProfile(
        complexity: Double,
        uncertainty: Double,
        isMultiHop: Boolean = false,
    ): QueryProfile {
        return QueryProfile(
            intent = QueryIntent.FACTUAL,
            complexity = complexity,
            uncertainty = uncertainty,
            requiredEvidence = setOf(EvidenceType.STRUCTURED_DATA),
            suggestedRoutes = listOf(Route.SQL),
            isMultiHop = isMultiHop,
            hasDependency = false,
        )
    }
}

File: agent-app-go/internal/core/model_escalator_test.go
package core

import (
	"testing"

	"github.com/riwonace/agent-app-go/internal/router"
)

func newEscalator() *ModelEscalator {
	return &ModelEscalator{Enabled: true, SmallModel: "gemma3:1b", MediumModel: "qwen2.5:3b", LargeModel: "qwen2.5:7b"}
}

func newProfile(complexity, uncertainty float64, isMultiHop bool) QueryProfile {
	return QueryProfile{
		Intent:           IntentFactual,
		Complexity:       complexity,
		Uncertainty:      uncertainty,
		RequiredEvidence: map[EvidenceType]bool{EvidenceStructuredData: true},
		SuggestedRoutes:  []router.Route{router.RouteSQL},
		IsMultiHop:       isMultiHop,
	}
}

func TestModelEscalator_LowComplexitySelectsSmall(t *testing.T) {
	sel := newEscalator().SelectModel(newProfile(0.2, 0.1, false))
	if sel.Tier != TierSmall || sel.Model != "gemma3:1b" {
		t.Fatalf("got %+v", sel)
	}
}

func TestModelEscalator_MediumComplexitySelectsMedium(t *testing.T) {
	sel := newEscalator().SelectModel(newProfile(0.5, 0.1, false))
	if sel.Tier != TierMedium || sel.Model != "qwen2.5:3b" {
		t.Fatalf("got %+v", sel)
	}
}

func TestModelEscalator_HighComplexitySelectsLarge(t *testing.T) {
	sel := newEscalator().SelectModel(newProfile(0.8, 0.1, false))
	if sel.Tier != TierLarge || sel.Model != "qwen2.5:7b" {
		t.Fatalf("got %+v", sel)
	}
}

func TestModelEscalator_HighUncertaintyBumpsOneTier(t *testing.T) {
	sel := newEscalator().SelectModel(newProfile(0.3, 0.6, false))
	if sel.Tier != TierMedium {
		t.Fatalf("got %+v", sel)
	}
}

func TestModelEscalator_MultiHopIsAtLeastMedium(t *testing.T) {
	sel := newEscalator().SelectModel(newProfile(0.2, 0.1, true))
	if sel.Tier != TierMedium {
		t.Fatalf("got %+v", sel)
	}
}

func TestModelEscalator_LowQualityRecommendsReescalation(t *testing.T) {
	current := ModelSelection{Model: "gemma3:1b", Tier: TierSmall}
	re := newEscalator().ShouldReescalate(current, 0.4, 0.5)
	if re == nil || re.Tier != TierMedium {
		t.Fatalf("got %+v", re)
	}
}

func TestModelEscalator_SufficientQualityDoesNotReescalate(t *testing.T) {
	current := ModelSelection{Model: "gemma3:1b", Tier: TierSmall}
	re := newEscalator().ShouldReescalate(current, 0.8, 0.9)
	if re != nil {
		t.Fatalf("expected nil, got %+v", re)
	}
}

func TestModelEscalator_AlreadyLargeDoesNotReescalate(t *testing.T) {
	current := ModelSelection{Model: "qwen2.5:7b", Tier: TierLarge}
	re := newEscalator().ShouldReescalate(current, 0.3, 0.3)
	if re != nil {
		t.Fatalf("expected nil, got %+v", re)
	}
}

func TestModelEscalator_DisabledAlwaysSelectsSmall(t *testing.T) {
	disabled := &ModelEscalator{Enabled: false, SmallModel: "gemma3:1b", MediumModel: "qwen2.5:3b", LargeModel: "qwen2.5:7b"}
	sel := disabled.SelectModel(newProfile(0.9, 0.9, false))
	if sel.Tier != TierSmall {
		t.Fatalf("got %+v", sel)
	}
}
